### PR TITLE
Bl metrics library

### PR DESCRIPTION
**Pull request by Runjing-Liu120**
Hi everyone, 

I've now added a `metrics.py` file containing some functions to evaluate metrics. 

In the `get_metrics` method of `SleepPhase`, we now return the TPR and PPV. I've removed those functions from `sdss.py`.  

In addition, we now evaluate the location/flux error slightly differently than before. Before, we were returning the average error in location/fluxes **only if** the number of estimated sources equals the number of true sources. That way, we could find a one-to-one and onto matching between the estimated and true sources; then for each match, we compare locations/fluxes. 

This was throwing errors on M2 because we never **exactly** get the true number of stars correctly (there are ~1000 on the image; we might estimate, for example 1012 stars instead of 1023 stars. Close, but not exactly right). 

Now, the evaluation of location/flux error can handle the situation where the true number and estimated number of sources differ. We do an "inner join" (in the spirit of the Dplyr operation) on the two lists of sources. We find the minimizing permutation to minimize the error in locations, but allow some sources to not have a match (so if there are more true stars than estimated stars, some true stars won't have a match; if there are more estimated stars than true stars, some estimated stars won't have a match). The matching is done with the Hungarian algorithm in Scipy. We then only evaluate the location/flux error on sources that were matched.

### CODE REVIEW

**Review by ZachariahPang — APPROVED** [review]

File: bliss/metrics.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+            est_mag = torch.log10(est_fluxes) * 2.5
+
+            # get TPR and PPV
+            tpr_i, ppv_i = get_tpr_ppv(true_locs, true_mag, est_locs, est_mag)
+
+            tpr_vec[i] = tpr_i
+            ppv_vec[i] = ppv_i
+
+            # get l1 error in locations and fluxes
+            locs_mae_i, fluxes_mae_i = get_l1_error(
```
Comment (ismael-mendoza) [inline bliss/metrics.py:146]: `fluxes_mae` could be a bit misleading since it's really magnitudes right? Maybe change the name to `mag_mae`?

**Review by ismael-mendoza — APPROVED** [review]
Looks really good @Runjing-Liu120 , thank you! I just had 1 small comment about a variable's name

**Comment by ismael-mendoza** [comment]
It would be excellent if @ZachariahPang  or @zzhaozhe-profolio re-ran their tests on m2 with this branch to see if #164 was resolved with Bryan's changes.

**Comment by ZachariahPang** [comment]
> It would be excellent if @ZachariahPang or @zzhaozhe-profolio re-ran their tests on m2 with this branch to see if #164 was resolved with Bryan's new metrics library (or in a separate PR, whatever is easier)

I run a quick `tune.py` and have the following result.
Screen Shot 2021-01-12 at 11 35 08 
I think the new metrics work well. We probably can close #164 upon merging this.

Something I noticed though, is that `tune.py` has a much less GPU load, around 30% most of the time comparing to high 90% previously. I think this makes sense since the Hungarian algorithm is on CPU which bottlenecks the tuning process. The `tune.py` running time is significantly longer. A quick fix for this might be returning metrics once a few epochs instead of every epoch. I can make those changes if you think that might be beneficial.

**Pull request merged**